[Backtracing] Add JSON backtrace output

al45tair opened the pull request:
Add an option to use JSON for the crash log output; this makes it easy to ingest crash logs using tools like Splunk or Elastic, which is extremely useful in server environments.

rdar://121430255

al45tair committed:
[Backtracing] Added JSON crash log option.

Also made it so the `sanitize` option causes the crash logs to not
include memory dumps.

Fixes #71057
rdar://121430255

al45tair committed:
[Backtracing] Fix memory capture output.

A couple of commas were in the wrong place.

al45tair committed:
[Backtracing] Add platform and architecture to JSON output.

The JSON output should include platform and architecture data.

rdar://121430255

al45tair committed:
[Docs] Update backtracing documentation.

Added documentation for the JSON output.

rdar://121430255

al45tair committed:
[Backtracing] Fix JSON output when not giving backtraces for all threads.

I'd omitted the key from the dictionary by mistake.

rdar://121430255

al45tair committed:
[Test][Backtracing] Added tests for JSON backtrace output.

Check that we generate valid JSON crash reports.

rdar://121430255

al45tair commented:
(The very first commit is being dealt with separately in #79007. But we need it for the JSON changes, because when in JSON mode we change the filename for the former feature.)

al45tair commented:
@swift-ci Please smoke test

czechboy0 commented on docs/Backtracing.rst:
```diff
@@ -330,3 +338,184 @@ large number of frames in a much smaller space than would otherwise be possible.
++-------------------+--------------------------------------------------------+
+| Field             | Value                                                  |
++===================+========================================================+
+| threads           | An array of thread record, one for each thread (if the |
```
```suggestion
| threads           | An array of thread records, one for each thread (if the |
```

czechboy0 reviewed (commented):
Looks great - caught one typo, and I have one question: should the JSON output be minified (at least have an option to do so?).

Some logging systems require that each JSON object is completely on a single line, so this prettified version might not work with them.

Even if you change to minified JSON, anyone at desk can run this through jq for reading, if needed.

al45tair committed:
Update docs/Backtracing.rst

Fix a typo.

Co-authored-by: Honza Dvorsky <honza@apple.com>

al45tair commented:
> Looks great - caught one typo, and I have one question: should the JSON output be minified (at least have an option to do so?).
> 
> Some logging systems require that each JSON object is completely on a single line, so this prettified version might not work with them.
>
> Even if you change to minified JSON, anyone at desk can run this through jq for reading, if needed.

The JSON isn't _strictly_ minified, since there is some whitespace, but there are (by design) no newlines. The tests are slightly misleading perhaps because they run the JSON through a JSON linter/prettifier, which means everything ends up on separate lines.

mikeash commented on docs/Backtracing.rst:
```diff
@@ -330,3 +338,184 @@ large number of frames in a much smaller space than would otherwise be possible.
+|                   |                                                        |
+|                   | The dictionary is keyed by architecture specific       |
+|                   | register name; values are given as hexadecimal         |
+|                   | strings.                                               |
```
Are these prefixed with `0x`? I assumed not reading here, but it looks like the code does emit the prefix.

mikeash commented on stdlib/public/libexec/swift-backtrace/Utils.swift:
```diff
@@ -173,7 +227,60 @@ struct CFileStream: TextOutputStream {
+
+/// Escape a JSON string
+func escapeJSON(_ s: String) -> String {
+  var result = ""
```
Not terribly important, but this might very slightly benefit from a `reserveCapacity`.

al45tair replied:
Seems a good idea.

mikeash reviewed (approved):
Looks good. A few suggestions, nothing I'd insist on strongly.

While reading the JSON output format, it occurs to me that it doesn't have the VM region info that we get in Mac crash logs. That can be really useful sometimes and it might be nice to add it here. That doesn't belong in this PR, but I thought I'd put that out there.

al45tair commented on docs/Backtracing.rst:
```diff
@@ -330,3 +338,184 @@ large number of frames in a much smaller space than would otherwise be possible.
+| threads           | An array of thread record, one for each thread (if the |
+|                   | backtracer is set to give backtraces for all threads). |
++-------------------+--------------------------------------------------------+
+| crashedThread     | A single thread record for the crashing thread (if the |
```
Good point. Perhaps I should have an `omittedThreads` counter, like I do for images. That would solve the problem nicely.

al45tair commented on docs/Backtracing.rst:
```diff
@@ -330,3 +338,184 @@ large number of frames in a much smaller space than would otherwise be possible.
+|                   | any.  This will not be present if the ``sanitize``     |
+|                   | setting is enabled, or if no data was captured.        |
+|                   |                                                        |
+|                   | The dictionary is keyed by hexadecimal addresses, as   |
```
The backtracer currently captures 16 bytes at valid addresses it finds in the registers. I haven't specified how much it captures here, more because I didn't want anyone to make assumptions about that.

al45tair replied:
I've updated the documentation slightly on this point.

al45tair commented on docs/Backtracing.rst:
```diff
@@ -330,3 +338,184 @@ large number of frames in a much smaller space than would otherwise be possible.
+|                   | backtracer is set to give a backtrace only for the     |
+|                   | crashed thread).                                       |
++-------------------+--------------------------------------------------------+
+| registers         | A dictionary containing the register contents on the   |
```
I think I was trying to make it mirror what the textual crash report does, but I think you're right, for JSON that does make more sense.

al45tair committed:
[Backtracing] Update JSON output slightly after review.

Always put `registers` in the thread record, and always use the `threads`
top level key, even if we ask for only the crashed thread.  Also add an
`omittedThreads` key.

rdar://121430255

al45tair committed:
[Backtracing] Use reserveCapacity to optimize escaping slightly.

We know we need at least as much space as the number of UTF-8
characters.

rdar://121430255

al45tair committed:
[Backtracing] Update copyright dates in boilerplate.

rdar://121430255

al45tair commented:
@swift-ci Please smoke test

al45tair committed:
[Backtracing] Fix a couple of Linux specific issues.

Linux gives the main thread a name based on the process name, and also
we need to tolerate an extra slash on the `<compiler-generated>` filename
for some reason.

rdar://121430255

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
Linux timed-out during an unrelated test.
